File: js/dist-cjs/orchestration/orchestrator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AgentOrchestrator = void 0;
const uuid_1 = require("uuid");
const pending_js_1 = require("./pending.js");
/**
 * AgentOrchestrator — wraps one or more AgentChatDelegate instances with:
 *   Phase 0: Pending confirmation handling
 *   Phase 1: Intent resolution
 *   Phase 2: Direct execution (bypass LLM)
 *   Phase 3: Delegate routing
 *
 * AgentOrchestrator itself satisfies AgentChatDelegate, enabling nesting.
 */
class AgentOrchestrator {
    config;
    constructor(config) {
        this.config = config;
    }
    async fireHook(fn) {
        if (fn) {
            try {
                await fn();
            }
            catch {
                // hooks do not affect main flow
            }
        }
    }
    async chat(message, options) {
        const startTime = Date.now();
        const traceId = (0, uuid_1.v4)();
        const sessionId = options?.sessionId ?? null;
        const timings = { total: 0 };
        let phaseReached = "delegate";
        let orchestrationIntent = null;
        let directExecutorUsed = null;
        let delegateUsed = null;
        // Fire beforeMessage hook
        await this.fireHook(this.config.lifecycleHooks?.beforeMessage
            ? () => this.config.lifecycleHooks.beforeMessage(message, options)
            : undefined);
        const buildDecisionTrace = (overrides) => ({
            traceId,
            phaseReached: overrides.phaseReached,
            intentResolved: overrides.orchestrationIntent,
            timings: overrides.timings,
            directExecutorUsed: overrides.directExecutorUsed,
            delegateUsed: overrides.delegateUsed,
        });
        const buildFinalResult = (baseResult, overrides) => {
            const decisionTrace = buildDecisionTrace(overrides);
            return {
                ...baseResult,
                content: overrides.content ?? baseResult.content,
                orchestrationIntent: overrides.orchestrationIntent,
                decisionTrace,
                pendingConfirmation: overrides.pendingConfirmation ?? null,
                sessionId: overrides.sessionId !== undefined ? overrides.sessionId : (sessionId ?? baseResult.sessionId),
            };
        };
        // Cache session state to avoid triple fetch within the same chat() call
        let cachedSessionState;
        try {
            // === Phase 0: Pending Confirmation ===
            if (this.config.pendingStateManager && sessionId) {
                const pending = await this.config.pendingStateManager.getPending(sessionId);
                if (pending) {
                    const classifier = this.config.confirmationClassifier ?? pending_js_1.classifyConfirmationDisposition;
                    const disposition = classifier(message);
                    if (disposition === "confirm" || disposition === "reject") {
                        // Handle confirmation/rejection
                        await this.config.pendingStateManager.clearPending(sessionId);
                        phaseReached = "pending_confirmation";
                        timings.total = Date.now() - startTime;
                        const confirmContent = disposition === "confirm"
                            ? `Confirmed: ${pending.type}`
                            : `Rejected: ${pending.type}`;
                        const baseResult = {
                            content: confirmContent,
                            blocked: false,
                            usage: { promptTokens: 0, completionTokens: 0, totalTokens: 0 },
                            intent: null,
                            toolCalls: [],
                            timings: { total: timings.total },
                            sessionId,
                            traceId,
                            trace: null,
                        };
                        const result = buildFinalResult(baseResult, {
                            phaseReached,
                            orchestrationIntent: null,
                            timings,
                            directExecutorUsed: null,
                            delegateUsed: null,
                            pendingConfirmation: null,
                        });
                        await this.fireHook(this.config.lifecycleHooks?.afterMessage
                            ? () => this.config.lifecycleHooks.afterMessage(result)
                            : undefined);
                        return result;
                    }
                    // override — clear pending, continue normal flow
                    await this.config.pendingStateManager.clearPending(sessionId);
                }
            }
            // === Phase 1: Intent Resolution ===
            if (this.config.intentResolver) {
                const intentStart = Date.now();
                if (this.config.sessionStateStore && sessionId && cachedSessionState === undefined) {
                    cachedSessionState = await this.config.sessionStateStore.get(sessionId);
                }
                orchestrationIntent = await this.config.intentResolver.resolve({
                    message,
                    sessionState: cachedSessionState ?? undefined,
                });
                timings.intentResolution = Date.now() - intentStart;
            }
            // === Phase 2: Direct Execution ===
            if (this.config.directExecutors && orchestrationIntent) {
                const execStart = Date.now();
                for (const executor of this.config.directExecutors) {
                    if (executor.canHandle(orchestrationIntent)) {
                        const execResult = await executor.execute({
                            message,
                            intent: orchestrationIntent,
                            options: options,
                        });
                        if (execResult !== null) {
                            phaseReached = "direct_execution";
                            directExecutorUsed = executor.name;
                            timings.directExecution = Date.now() - execStart;
                            timings.total = Date.now() - startTime;
                            const result = buildFinalResult(execResult, {
                                phaseReached,
                                orchestrationIntent,
                                timings,
                                directExecutorUsed,
                                delegateUsed: null,
                            });
                            await this.fireHook(this.config.lifecycleHooks?.afterMessage
                                ? () => this.config.lifecycleHooks.afterMessage(result)
                                : undefined);
                            return result;
                        }
                    }
                }
            }
            // === Phase 3: Delegate Routing ===
            const delegateStart = Date.now();
            let selectedDelegate = this.config.delegate;
            let selectedDelegateName = "default";
            if (orchestrationIntent && this.config.delegates) {
                const mapped = this.config.delegates.get(orchestrationIntent.object);
                if (mapped) {
                    selectedDelegate = mapped;
                    selectedDelegateName = orchestrationIntent.object;
                }
            }
            // Enrich options with session state if available (use cached fetch)
            let enrichedOptions = options;
            if (this.config.sessionStateStore && sessionId) {
                if (cachedSessionState === undefined) {
                    cachedSessionState = await this.config.sessionStateStore.get(sessionId);
                }
                if (cachedSessionState) {
                    enrichedOptions = { ...options, sessionState: cachedSessionState };
                }
            }
            const delegateResult = await selectedDelegate.chat(message, enrichedOptions);
            timings.delegate = Date.now() - delegateStart;
            timings.total = Date.now() - startTime;
            phaseReached = "delegate";
            delegateUsed = selectedDelegateName;
            // Update session state if applicable (reuse cached session state)
            if (this.config.sessionStateStore && sessionId && delegateResult.sessionId) {
                const newState = {
                    sessionId,
                    lastPresentedEntities: cachedSessionState?.lastPresentedEntities ?? [],
                    metadata: cachedSessionState?.metadata ?? {},
                    updatedAt: Date.now(),
                };
                await this.config.sessionStateStore.save(sessionId, newState);
            }
            const result = buildFinalResult(delegateResult, {
                phaseReached,
                orchestrationIntent,
                timings,
                directExecutorUsed,
                delegateUsed,
                sessionId: sessionId ?? delegateResult.sessionId,
            });
            await this.fireHook(this.config.lifecycleHooks?.afterMessage
                ? () => this.config.lifecycleHooks.afterMessage(result)
                : undefined);
            return result;
        }
        catch (error) {
            // Fire onError hook
            await this.fireHook(this.config.lifecycleHooks?.onError
                ? () => this.config.lifecycleHooks.onError(error)
                : undefined);
            // Re-throw after hook
            throw error;
        }
    }
    /**
     * Process a channel-specific raw input through the channel adapter.
     */
    async processChannel(rawInput) {
        if (!this.config.channelAdapter) {
            throw new Error("No channelAdapter configured");
        }
        const channelMessage = this.config.channelAdapter.parseIncoming(rawInput);
        const result = await this.chat(channelMessage.text, channelMessage.options);
        return this.config.channelAdapter.formatOutgoing(result);
    }
}
exports.AgentOrchestrator = AgentOrchestrator;

File: naru_agent/orchestration/orchestrator.py
"""AgentOrchestrator — 4-phase routing + multi-agent orchestration."""

from __future__ import annotations

import logging
import time
import uuid
from dataclasses import dataclass, field
from typing import Any, Callable, Generic, Protocol, TypeVar

from naru_agent.agent import NaruResult
from naru_agent.orchestration.channel import ChannelAdapter, ChannelMessage
from naru_agent.orchestration.executor import BaseDirectExecutor
from naru_agent.orchestration.intent import (
    BaseIntentResolver,
    IntentResolveInput,
    OrchestratorIntent,
)
from naru_agent.orchestration.pending import (
    BasePendingStateManager,
    PendingState,
    classify_confirmation_disposition,
)
from naru_agent.orchestration.result import OrchestrationResult, PendingConfirmation
from naru_agent.orchestration.session_state import AgentSessionState, BaseSessionStateStore
from naru_agent.orchestration.trace import AgentDecisionTrace, OrchestrationPhase, OrchestrationTimings

logger = logging.getLogger(__name__)

T = TypeVar("T", bound=str)


class AgentChatDelegate(Protocol):
    """Any object with a chat() method.

    NaruAgent satisfies this structurally (duck typing / PEP 544).
    """

    def chat(
        self,
        message: str,
        user_id: str | None = None,
        session_id: str | None = None,
    ) -> NaruResult:
        ...


@dataclass
class LifecycleHooks:
    """Optional lifecycle hooks — exceptions are swallowed to protect main flow."""

    before_message: Callable[[str, str | None, str | None], None] | None = None
    after_message: Callable[[OrchestrationResult], None] | None = None
    on_error: Callable[[BaseException], None] | None = None


@dataclass
class AgentOrchestratorConfig(Generic[T]):
    """Configuration for AgentOrchestrator."""

    delegate: AgentChatDelegate
    delegates: dict[T, AgentChatDelegate] = field(default_factory=dict)
    intent_resolver: BaseIntentResolver[T] | None = None
    direct_executors: list[BaseDirectExecutor[T]] = field(default_factory=list)
    pending_state_manager: BasePendingStateManager | None = None
    session_state_store: BaseSessionStateStore | None = None
    channel_adapter: ChannelAdapter | None = None
    lifecycle_hooks: LifecycleHooks | None = None
    confirmation_classifier: Callable[[str], str] | None = None


class AgentOrchestrator(Generic[T]):
    """Wraps one or more AgentChatDelegate instances with 4-phase routing.

    Phase 0: Pending confirmation handling
    Phase 1: Intent resolution
    Phase 2: Direct execution (bypass LLM)
    Phase 3: Delegate routing

    AgentOrchestrator itself satisfies AgentChatDelegate, enabling nesting.
    """

    def __init__(self, config: AgentOrchestratorConfig[T]) -> None:
        self._config = config

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _fire_hook(self, fn: Callable | None, *args: Any) -> None:
        """Call a lifecycle hook, silently swallowing any exception."""
        if fn is None:
            return
        try:
            fn(*args)
        except Exception:  # noqa: BLE001
            pass

    def _wrap_result(
        self,
        base_result: NaruResult,
        *,
        trace_id: str,
        phase_reached: OrchestrationPhase,
        orchestration_intent: OrchestratorIntent[T] | None,
        timings: OrchestrationTimings,
        direct_executor_used: str | None = None,
        delegate_used: str | None = None,
        pending_confirmation: PendingConfirmation | None = None,
        override_session_id: str | None = None,
    ) -> OrchestrationResult:
        """Wrap a NaruResult with orchestration metadata."""
        decision_trace = AgentDecisionTrace(
            trace_id=trace_id,
            phase_reached=phase_reached,
            intent_resolved=orchestration_intent,
            timings=timings,
            direct_executor_used=direct_executor_used,
            delegate_used=delegate_used,
        )
        return OrchestrationResult(
            content=base_result.content,
            blocked=base_result.blocked,
            usage=base_result.usage,
            intent=base_result.intent,
            tool_calls=base_result.tool_calls,
            timings=base_result.timings,
            session_id=override_session_id if override_session_id is not None else base_result.session_id,
            trace_id=base_result.trace_id,
            trace=base_result.trace,
            decision_trace=decision_trace,
            pending_confirmation=pending_confirmation,
            orchestration_intent=orchestration_intent,
        )

    def _get_hook(self, name: str) -> Callable | None:
        """Return a lifecycle hook by name, or None."""
        hooks = self._config.lifecycle_hooks
        if hooks is None:
            return None
        return getattr(hooks, name, None)

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def chat(
        self,
        message: str,
        user_id: str | None = None,
        session_id: str | None = None,
    ) -> OrchestrationResult:
        """Route a message through the 4-phase orchestration pipeline."""
        start_time = time.monotonic()
        trace_id = str(uuid.uuid4())

        timings = OrchestrationTimings()
        orchestration_intent: OrchestratorIntent[T] | None = None

        self._fire_hook(self._get_hook("before_message"), message, user_id, session_id)

        # Cache session state to avoid repeated fetches
        cached_session_state: AgentSessionState | None = None
        session_state_fetched = False

        try:
            # === Phase 0: Pending Confirmation ===
            if self._config.pending_state_manager and session_id:
                pending = self._config.pending_state_manager.get_pending(session_id)
                if pending is not None:
                    classifier = (
                        self._config.confirmation_classifier
                        or classify_confirmation_disposition
                    )
                    disposition = classifier(message)

                    if disposition in ("confirm", "reject"):
                        self._config.pending_state_manager.clear_pending(session_id)
                        timings.total = time.monotonic() - start_time
                        content = (
                            f"Confirmed: {pending.type}"
                            if disposition == "confirm"
                            else f"Rejected: {pending.type}"
                        )
                        base_result = NaruResult(
                            content=content,
                            blocked=False,
                            usage={"prompt_tokens": 0, "completion_tokens": 0, "total_tokens": 0},
                            timings={"total": timings.total},
                            session_id=session_id,
                            trace_id=trace_id,
                        )
                        result = self._wrap_result(
                            base_result,
                            trace_id=trace_id,
                            phase_reached="pending_confirmation",
                            orchestration_intent=None,
                            timings=timings,
                        )
                        self._fire_hook(self._get_hook("after_message"), result)
                        return result

                    # override — clear pending, continue normal flow
                    self._config.pending_state_manager.clear_pending(session_id)

            # === Phase 1: Intent Resolution ===
            if self._config.intent_resolver:
                intent_start = time.monotonic()
                if self._config.session_state_store and session_id and not session_state_fetched:
                    cached_session_state = self._config.session_state_store.get(session_id)
                    session_state_fetched = True

                orchestration_intent = self._config.intent_resolver.resolve(
                    IntentResolveInput(
                        message=message,
                        session_state=cached_session_state,
                    )
                )
                timings.intent_resolution = time.monotonic() - intent_start

            # === Phase 2: Direct Execution ===
            if self._config.direct_executors and orchestration_intent is not None:
                exec_start = time.monotonic()
                for executor in self._config.direct_executors:
                    if executor.can_handle(orchestration_intent):
                        exec_result = executor.execute(
                            message=message,
                            intent=orchestration_intent,
                            options={
                                "user_id": user_id,
                                "session_id": session_id,
                                "session_state": cached_session_state,
                            },
                        )
                        if exec_result is not None:
                            timings.direct_execution = time.monotonic() - exec_start
                            timings.total = time.monotonic() - start_time

                            result = self._wrap_result(
                                exec_result,
                                trace_id=trace_id,
                                phase_reached="direct_execution",
                                orchestration_intent=orchestration_intent,
                                timings=timings,
                                direct_executor_used=executor.name,
                            )
                            self._fire_hook(self._get_hook("after_message"), result)
                            return result

            # === Phase 3: Delegate Routing ===
            delegate_start = time.monotonic()
            selected_delegate: AgentChatDelegate = self._config.delegate
            selected_delegate_name = "default"

            if orchestration_intent is not None and self._config.delegates:
                mapped = self._config.delegates.get(orchestration_intent.object)  # type: ignore[arg-type]
                if mapped is not None:
                    selected_delegate = mapped
                    selected_delegate_name = str(orchestration_intent.object)

            # Fetch session state if not yet done (for post-delegate save)
            # NOTE: JS version injects sessionState into enrichedOptions passed
            # to the delegate. Python's AgentChatDelegate Protocol has a fixed
            # signature (message, user_id, session_id) matching NaruAgent.chat(),
            # so session state cannot be forwarded as a parameter. Delegates that
            # need session state should read from the store directly.
            if self._config.session_state_store and session_id and not session_state_fetched:
                cached_session_state = self._config.session_state_store.get(session_id)
                session_state_fetched = True

            delegate_result = selected_delegate.chat(
                message,
                user_id=user_id,
                session_id=session_id,
            )
            timings.delegate = time.monotonic() - delegate_start
            timings.total = time.monotonic() - start_time

            # Update session state if applicable
            if self._config.session_state_store and session_id and delegate_result.session_id:
                new_state = AgentSessionState(
                    session_id=session_id,
                    last_presented_entities=cached_session_state.last_presented_entities
                    if cached_session_state
                    else [],
                    metadata=cached_session_state.metadata if cached_session_state else {},
                    updated_at=time.time(),
                )
                self._config.session_state_store.save(session_id, new_state)

            result = self._wrap_result(
                delegate_result,
                trace_id=trace_id,
                phase_reached="delegate",
                orchestration_intent=orchestration_intent,
                timings=timings,
                delegate_used=selected_delegate_name,
                override_session_id=session_id or delegate_result.session_id,
            )
            self._fire_hook(self._get_hook("after_message"), result)
            return result

        except Exception as exc:
            self._fire_hook(self._get_hook("on_error"), exc)
            raise

    def process_channel(self, raw_input: Any) -> Any:
        """Process channel-specific raw input through the configured channel adapter."""
        if self._config.channel_adapter is None:
            raise ValueError("No channel_adapter configured")

        channel_message: ChannelMessage = self._config.channel_adapter.parse_incoming(raw_input)

        result = self.chat(
            channel_message.text,
            user_id=channel_message.user_id,
            session_id=channel_message.session_id,
        )

        return self._config.channel_adapter.format_outgoing(result)
